packages/orama/src/components/index.ts

```diff
@@ -208,32 +210,60 @@ export async function create(
   return index
 }
 
-export async function insert(
+async function insertScalar(
   implementation: IIndex<Index>,
   index: Index,
   prop: string,
   id: string,
   value: SearchableValue,
+  schemaType: SearchableType,
   language: string | undefined,
   tokenizer: Tokenizer,
   docsCount: number,
 ): Promise<void> {
-  if (typeof value === 'number') {
-    avlInsert(index.indexes[prop] as AVLNode<number, string[]>, value as number, [id])
-    return
-  } else if (typeof value === 'boolean') {
-    // eslint-disable-next-line @typescript-eslint/no-extra-semi
-    ;(index.indexes[prop] as BooleanIndex)[value ? 'true' : 'false'].push(id)
-    return
+  switch (schemaType) {
+    case 'boolean':
+      (index.indexes[prop] as BooleanIndex)[value ? 'true' : 'false'].push(id)
+      break
+    case 'number':
+      avlInsert(index.indexes[prop] as AVLNode<number, string[]>, value as number, [id])
+      break
+    case 'string': {
+      const tokens = await tokenizer.tokenize(value as string, language, prop)
+      await implementation.insertDocumentScoreParameters(index, prop, id, tokens, docsCount)
+
+      for (const token of tokens) {
+        await implementation.insertTokenScoreParameters(index, prop, id, tokens, token)
+
+        radixInsert(index.indexes[prop] as RadixNode, token, id)
+      }
+
+      break
+    }
   }
+}
 
-  const tokens = await tokenizer.tokenize(value as string, language, prop)
+export async function insert(
+  implementation: IIndex<Index>,
+  index: Index,
+  prop: string,
+  id: string,
+  value: SearchableValue,
+  schemaType: SearchableType,
+  language: string | undefined,
+  tokenizer: Tokenizer,
+  docsCount: number,
+): Promise<void> {
+  if (!schemaType.endsWith('[]')) {
```

Comment:
Are we sure we don't fall into the `foo[]` error here?

Reply:
If I'm not mistaken, it would be rejected in `create` phase.

Reply:
Correct, during the creation phase, the schema is validated. Also, `schemaType` is typed. So at compile-time and run-time, we are covered.